core-foundation/src/dictionary.rs

```diff
@@ -130,4 +249,133 @@ impl CFDictionary {
+    }
+
+    /// Similar to `set` but acts on a higher level, automatically converting from any `TCFType`
+    /// to the raw pointer of its concrete TypeRef.
+    #[inline]
+    pub fn set2<RK, RV, K, V>(&self, key: &K, value: &V)
+        where K: TCFType<*const RK>,
+              V: TCFType<*const RV> {
+        unsafe {
+            self.set(key.as_concrete_TypeRef() as *const c_void,
+                     value.as_concrete_TypeRef() as *const c_void)
+        }
+    }
+
+    /// Replaces the value of the key in the dictionary.
+    #[inline]
+    pub unsafe fn replace(&self, key: *const c_void, value: *const c_void) {
+        CFDictionaryReplaceValue(self.0, key, value)
+    }
+
+    /// Similar to `replace` but acts on a higher level, automatically converting from any `TCFType`
+    /// to the raw pointer of its concrete TypeRef.
+    #[inline]
+    pub fn replace2<RK, RV, K, V>(&self, key: &K, value: &V)
+        where K: TCFType<*const RK>,
+              V: TCFType<*const RV> {
+        unsafe {
+            self.replace(key.as_concrete_TypeRef() as *const c_void,
+                         value.as_concrete_TypeRef() as *const c_void)
+        }
+    }
+
+    /// Removes the value of the key from the dictionary.
+    #[inline]
+    pub unsafe fn remove(&self, key: *const c_void) {
+        CFDictionaryRemoveValue(self.0, key);
+    }
+
+    /// Similar to `remove` but acts on a higher level, automatically converting from any `TCFType`
+    /// to the raw pointer of its concrete TypeRef.
+    #[inline]
+    pub fn remove2<RK, K>(&self, key: &K)
+        where K: TCFType<*const RK> {
+        unsafe {
+            self.remove(key.as_concrete_TypeRef() as *const c_void)
+        }
+    }
+
+    #[inline]
+    pub fn remove_all(&self) {
+        unsafe { CFDictionaryRemoveAllValues(self.0) }
+    }
 }
+
+// This is the contents of the impl_TCFType! macro so I can implement Clone trait.
+impl TCFType<CFMutableDictionaryRef> for CFMutableDictionary {
+    #[inline]
+    fn as_concrete_TypeRef(&self) -> CFMutableDictionaryRef {
+        self.0
+    }
+
+    #[inline]
+    unsafe fn wrap_under_get_rule(reference: CFMutableDictionaryRef) -> CFMutableDictionary {
+        let reference = ::std::mem::transmute(::core_foundation_sys::base::CFRetain(::std::mem::transmute(reference)));
+        TCFType::wrap_under_create_rule(reference)
+    }
+
+    #[inline]
+    fn as_CFTypeRef(&self) -> ::core_foundation_sys::base::CFTypeRef {
+        unsafe {
+            ::std::mem::transmute(self.as_concrete_TypeRef())
+        }
+    }
+
+    #[inline]
+    unsafe fn wrap_under_create_rule(obj: CFMutableDictionaryRef) -> CFMutableDictionary {
+        CFMutableDictionary(obj)
+    }
+
+    #[inline]
+    fn type_id() -> ::core_foundation_sys::base::CFTypeID {
+        unsafe {
+            CFDictionaryGetTypeID()
+        }
+    }
+}
+
+impl Clone for CFMutableDictionary {
+    #[inline]
+    fn clone(&self) -> CFMutableDictionary {
+        unsafe {
```

Comment:
self.copy_with_capacity?

Reply:
Thanks for the review, @jdm. I pushed the requested changes.